### PR TITLE
[SQL] Improve error messages

### PR DESCRIPTION
**Pull request by marmbrus**

### CODE REVIEW

**Comment by rxin** [comment]
This is going to be huge. You probably want a JIRA ticket for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -66,32 +66,81 @@ class Analyzer(catalog: Catalog,
...
-      } match {
-        // As a backstop, use the root node to check that the entire plan tree is resolved.
-        case p if !p.resolved =>
-          throw new AnalysisException(s"Unresolved operator in the query plan ${p.simpleString}")
-        case p => p
+      // We transform up and order the rules so as to catch the first possible failure instead
+      // of the result of cascading resolution failures.
+      plan.foreachUp {
+        case operator: LogicalPlan =>
+          operator transformAllExpressions {
```
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: we can use transformExpressionsUp here to cover the case of ``` SELECT CAST(x AS STRING) FROM src```

**Comment by scwf** [comment]
@marmbrus, i filed a Pr for your branch(https://github.com/marmbrus/spark/pull/15), hopefully it is useful

**Comment by marmbrus** [comment]
Thanks for the PR!

**Pull request closed without merging**